UsbIpServer/RegistryUtils.cs

```diff
@@ -5,45 +5,207 @@
 using System;
 using System.Linq;
 using System.Security.Principal;
+using System.Globalization;
 using Microsoft.Win32;
+using System.Collections.Generic;
 
 namespace UsbIpServer
 {
     static class RegistryUtils
     {
         const string DevicesRegistryPath = @"SOFTWARE\usbipd-win";
 
-        public static bool IsDeviceAvailable(string busId)
+        static class DeviceFilter
         {
-            return Registry.LocalMachine.OpenSubKey(DevicesRegistryPath)?.GetSubKeyNames().Any(x => x == busId) ?? false;
+            public const string VENDOR_ID = "VendorId";
+            public const string PRODUCT_ID = "ProductId";
+            public const string BCD_DEVICE = "BcdDevice";
+            public const string DEVICE_CLASS = "DeviceClass";
+            public const string DEVICE_SUBCLASS = "DeviceSubClass";
+            public const string DEVICE_PROTOCOL = "DeviceProtocol";
+            public const string DEV_NUM = "DevNum";
+            public const string BUS_ID = "BusId";
         }
 
-        public static void SetDeviceAvailability(string busId, bool enable)
+        public static bool IsDeviceShared(ExportedDevice device)
         {
-            if (enable)
-            { 
-                Registry.LocalMachine.CreateSubKey($@"{DevicesRegistryPath}\{busId}");
-            } else if (IsDeviceAvailable(busId))
+            return GetRegistryKey(device) != null;
+        }
+
+        public static bool IsDeviceAttached(ExportedDevice device)
+        {
+            var key = GetRegistryKey(device);
+            if (key != null)
+            {
+                return key.GetSubKeyNames().Contains("Attached");
```

Comment:
Maybe nitpicking here, but when you kill the server (not a proper shutdown), the registry key is not removed (of course). That's OK, but this will list the device as Attached anyway. Maybe this should be && with service running status. In fact, maybe it would be nice to even display a warning at the end of the list if the service is not running (which it normally should always be).

Reply:
That's a good way to handle this. I'll make the fix!

Reply:
Done